sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -1023,26 +957,26 @@ int gcoap_get_resource_list(void *buf, size_t maxlen, uint8_t cf)
 
 int gcoap_add_qstring(coap_pkt_t *pdu, const char *key, const char *val)
 {
-    size_t qs_len = strlen((char *)pdu->qs);
+    char qs[NANOCOAP_QS_MAX];
+    size_t qs_len  = 0;
     size_t key_len = strlen(key);
     size_t val_len = (val) ? (strlen(val) + 1) : 0;
 
-    /* make sure if url_len + the new query string fit into the url buffer */
-    if ((qs_len + key_len + val_len + 2) >= NANOCOAP_QS_MAX) {
+    /* add 2 for '=' and null terminator */
+    if ((key_len + val_len + 2) >= NANOCOAP_QS_MAX) {
```

Comment:
the `+ 2` seems wrong, as the `=` character is already accounted for when computing `val_len` above, right?

I propose the following code for this function:
```c
int gcoap_add_qstring(coap_pkt_t *pdu, const char *key, const char *val)
{
    char qs[NANOCOAP_QS_MAX];
    size_t len = strlen(key);
    size_t val_len = (val) ? (strlen(val) + 1) : 0;

    /* test if the query string fits, account for the zero termination */
    if ((len + val_len + 1) >= NANOCOAP_QS_MAX) {
        return -1;
    }

    memcpy(&qs[0], key, len);
    if (val) {
        qs[len] = '=';
        /* the `=` character was already counted in `val_len`, so subtract it here */
        memcpy(&qs[len + 1], val, (val_len - 1));
        len += val_len;
    }
    qs[len] = '\0';

    return coap_opt_add_string(pdu, COAP_OPT_URI_QUERY, qs, '&');
}
```
Its slightly shorter (though the compiled binary has the same size)... I'd say up to you.

Reply:
Yes, you're right about `+ 2`. I misinterpreted the reason for the addition in the old code. It was for the '&' and the null terminator, not for the '=' and the null. However, we no longer need to consider the '&' for the local `qs` buffer. So, the test should be changed to `+ 1`.

Thanks for the suggested replacement. I'll review.

Reply:
I'll incorporate your code. I agree that yours is an improvement and cleans up the memcpy and length accounting for the value. It looks like we were writing two null terminators in the old code.

At the same time, I'll add a couple of unit tests -- with and without a value. The nature of the code and the amount of change in this reimplementation really call for it.

Reply:
BTW, once this PR is merged, this function definitely should be reimplemented in nanocoap as coap_add_qstring(). It's no longer specific to gcoap. We can retain an inline for the existing function name.

Reply:
Reimplemented gcoap_add_string() in 33210ac, and added test in 0e275f9.

Reply:
> BTW, once this PR is merged, this function definitely should be reimplemented in nanocoap as coap_add_qstring() or something like that. It's no longer specific to gcoap. We can retain an inline for the existing function name.

+1!